### DISCUSSION TITLE
Debug ATMega32U4

### DISCUSSION THREAD

**Discussion by LaCebraDeAndalucia** [felias-fogg/dw-link#15] [Q&A]
Hello,
For a beginner who's just starting out... I'm expecting delivery of my first board in a week's time.
I'm finding conflicting advice on whether I can use dw-link (on an Arduino UNO board as a probe) to debug an Arduino Leonardo (ATmega32U4).
So debugging an ATmega32U4: Yes or No?
Thanks for sharing your experience.

**Comment by felias-fogg** [comment]
Hi,

In principle, you should be able to debug an ATmega32U4 with dw-link. However, I never tried it out yet. So there is a significant chance that something does not work out. And there are a number of potential problems if you want to debug while there is a USB connection to the ATmega32U4:

1. USB works only if the USB interrupts are serviced at regular times. If execution is stopped, then the USB connection is dropped. This means, you could use the debugger to run to a breakpoint and inspect the internal state. However, it is probably not possible to continue executing from this point, and you need to start over.
2. If you use a USB connection to the ATmega32U4, you either have to make sure that there is no power connection through the USB connector (by using a special [USB power blocker](https://www.amazon.de/PortaPow-USB-Power-Blocker/dp/B094FYL9QT/?th=1)), or you need to do the power-cycling that is necessary to enable debugWIRE mode manually.
3. In order to debug the chip, the entire chip (including the bootloader) has to be erased. This means that after debugging, you need to reflash the bootloader. 

All in all, with only a week's time, you probably should refrain from using dw-link. If you nevertheless try, I will definitely be interested in the outcome!  BTW: I am currently preparing board manager files that will finally allow debugging in Arduino IDE. 

Best regards,
Bernhard

**Comment by LaCebraDeAndalucia** [comment]
Thank you for your long and detailed reply. I'll take it into account gradually.

1/ I've since received two Leonardo cards, and things didn't go as planned 🙁
The first card worked for about ten minutes. I was able to transfer a program to it (only once). When I put my hand close to the card, its LEDs were flashing in disorder. I thought that the level of some inputs was not defined and that this explained it. But no, I couldn't make contact with the card again via the USB port.
I then used the second card, but on another computer, just to be on the safe side. It worked for two days, then stopped when I disconnected the USB cable (and used a USB key on the computer for a few moments instead).
I had also ordered two SparkFun programmers. With neither of them did I manage to re-establish contact with either of the Leonardo cards 🙁 🙁 🙁

2/ I should receive an Arduino UNO board tomorrow (the original from Arduino). I took the oldest model, it's very vintage with a USB type B socket and a µC in a DIL box!
So I won't be testing dw-link with Leonardo in the short term, sorry for you! 🙁

I'm going to resume my technical refresher (I've had a very long break...) with this UNO board before going further into debugging.
After that, I'm thinking of using another simple Arduino Nano board for :
- ISP flashing tools, I want to try something other than USB flashing;
- probe to try dw-link ;
- to understand the difficulties of porting the same code to two different µCs (UNO and NANO), without using the facilities provided by IDE 2 to do it automatically. This will lead me to master the libraries that interface two different µCs 🙃🙃🙃
Question: AM I ON THE RIGHT TRACK BY COMBINING UNO (as main) AND NANO (as auxiliary)?

3/ Finally, I'll be able to focus on the goal that motivates me: using three-axis accelerometers to record seismic activity in my region.
I hear it's active all the time. I once experienced a level 5 earthquake, no damage for me, but it was very noisy.

I'll let you know later, as I progress or as I have questions.
See you soon.
Fabrice
(translated from French into English thanks to DeepL)

________________________________
De : Bernhard ***@***.***>
Envoyé : samedi 28 décembre 2024 7:37
À : felias-fogg/dw-link ***@***.***>
Cc : LaCebraDeAndalucia ***@***.***>; Author ***@***.***>
Objet : Re: [felias-fogg/dw-link] Debug ATMega32U4 (Discussion #15)


Hi,

In principle, you should be able to debug an ATmega32U4 with dw-link. However, I never tried it out yet. So there is a significant chance that something does not work out. And there are a number of potential problems if you want to debug while there is a USB connection to the ATmega32U4:

  1.  USB works only if the USB interrupts are serviced at regular times. If execution is stopped, then the USB connection is dropped. This means, you could use the debugger to run to a breakpoint and inspect the internal state. However, it is probably not possible to continue executing from this point, and you need to start over.
  2.  If you use a USB connection to the ATmega32U4, you either have to make sure that there is no power connection through the USB connector (by using a special USB power blocker<https://www.amazon.de/PortaPow-USB-Power-Blocker/dp/B094FYL9QT/?th=1>), or you need to do the power-cycling that is necessary to enable debugWIRE mode manually.
  3.  In order to debug the chip, the entire chip (including the bootloader) has to be erased. This means that after debugging, you need to reflash the bootloader.

All in all, with only a week's time, you probably should refrain from using dw-link. If you nevertheless try, I will definitely be interested in the outcome! BTW: I am currently preparing board manager files that will finally allow debugging in Arduino IDE.

Best regards,
Bernhard

**Reply by felias-fogg** [reply]
Hi, 

I was actually too fast when I answered your first question. The problem is much  more severe. The AtmegaXU4 MCUs do not support debugWIRE. So, you only could use the Microchip tools. But there, you might face the issues I mentioned in my first response.

Cheers
